test/tests/functional/pbs_cgroups_hook.py

```diff
@@ -4334,6 +4476,274 @@ def test_cgroup_default_config(self):
+
+            cgswap = None
+            job_status = self.server.status(JOB, id=jid)
+            select_resource = job_status[0]['Resource_List.select']
+            chunkspecs = select_resource.split(':')
+            for c in chunkspecs:
+                if '=' in c:
+                    name, value = c.split('=')
+                    if name == 'cgswap':
+                        cgswap = PbsTypeSize(value)
+            self.assertTrue(cgswap is not None, 'job cgswap was not added')
+            self.assertTrue(cgswap == PbsTypeSize('1000mb'),
+                            'job cgswap is %s instead of expected 1000mb'
+                            % str(cgswap))
+            self.logger.info('job cgswap detected to be correct, roughly %s'
+                             % str(cgswap))
+        else:
+            self.logger.info('First MoM has swap, confirming cgswap '
+                             'correctly throttles jobs accepted')
+            # PbsTypeSize value is stored in kb units
+            cgreqval = int(float(cgswap.value)
+                           / 1024.0 / 3.0 * 2.0)
+            cgreqsuffix = 'mb'
+            cgreq = PbsTypeSize(str(cgreqval) + cgreqsuffix)
+            vmemreqsize = PbsTypeSize("100mb") + cgreq
+            vmemreq = str(int(vmemreqsize.value / 1024))+'mb'
+            self.logger.info('will submit jobs with 100mb mem and %s vmem'
+                             % vmemreq)
+            a = {'Resource_List.select':
+                 '1:ncpus=1:mem=100mb:vmem=%s:vnode=%s'
+                 % (vmemreq, vnode_name)}
+
+            j = Job(TEST_USER, attrs=a)
+            j.create_script(self.sleep100_job)
+            jid = self.server.submit(j)
+            bs = {'job_state': 'R'}
+            self.server.expect(JOB, bs, jid, offset=1)
+
+            cgswap = None
+            job_status = self.server.status(JOB, id=jid)
+            select_resource = job_status[0]['Resource_List.select']
+            chunkspecs = select_resource.split(':')
+            for c in chunkspecs:
+                if '=' in c:
+                    name, value = c.split('=')
+                    if name == 'cgswap':
+                        cgswap = PbsTypeSize(value)
+            self.assertTrue(cgswap is not None, 'job cgswap was not added')
+            self.assertTrue(cgswap == cgreq,
+                            'job cgswap is %s instead of expected %s'
+                            % (str(cgswap), str(cgreq)))
+            self.logger.info('job cgswap detected to be correct, roughly %s'
+                             % str(cgswap))
+            j = Job(TEST_USER, attrs=a)
+            j.create_script(self.sleep100_job)
+            jid = self.server.submit(j)
+
+            # Second job should not run - not enough cgswap
+            # offset large enough to give scheduler a chance
+            bs2 = {'job_state': 'Q'}
+            self.server.expect(JOB, bs2, jid, offset=20, max_attempts=1)
```

Comment:
Rather than wait for a scheduling cycle.  Set scheduling=t to cause a scheduling cycle to run.
And get rid off the offset.

Reply:
There is only ONE attempt -- the offset is necessary or it won't work.

As said before, we don't need to "trigger a cycle" -- a cycle *is* triggered. The problem is that on a test harness with variable performance and variable times to actually run a cycle to completion, we do not know exactly how long to wait before the scheduler has had a chance to run and MoM has had a chance to accept the job and tell the server. The 20 seconds is safe on a well-known test harness. Less, and you're going to run the chance that you're not seeing the job rujn simply because the test harness is too slow toi actually run the job, even though it (mistakenly) does.

 This works, and if you want to propose an alternative, I'm open to suggestions, but one proven to work on a system where operations that could take less than a second on a "regular" host would take 10 seconds.

Reply:
Note: I am going to increase a sleep in the setup routine from 4 seconds to 6 seconds, BTW. A MoM bug together with the extreme slowness of the tests harness VMs has shown once case where "only" 4 seconds has introduced a false failure because when waiting 4 seconds, we were sending a HUP signal just when MoM was setting up the TPP thread, which triggered a race condition and a hang.

And yes: after 4 whole seconds the MoM wasn't fully restarted, and had not even begun running the exechost_startup hooks.

Reply:
How about:
1) submit job 1
2) expect job 1 to be in R state
3) submit job 2
4) call run_scheduling_cycle 
5) expect job 2 in Q state